### ISSUE TITLE
any perl versions core dumps with this code

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#5739]
Migrated from [rt.perl.org#15005](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=15005) (status was 'resolved')

Searchable as RT15005$

**Comment by p5pRT** [comment]
### From sem@ciam.ru

#### Created by sem@ciam.ru


This code always make perl core dump&#8203;:
\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*
my %level\_warnings;
sub stat\_level\_warning
\{
&nbsp;    sub \($level\, $message\) = @&#8203;\_;
&nbsp;    if\( $level\_warnings\{$message\} eq undef \)
&nbsp;    \{
&nbsp;    $level\_warnings\{$message\} = \[\];
&nbsp;    for my $l \( 0 \.\. $levels \)
&nbsp;    \{
&nbsp;        $level\_warnings\{$message\}\->\[$l\] = 0;
&nbsp;    \}
&nbsp;    \}
&nbsp;    $level\_warnings\{$message\}\->\[$level\]\+\+;
\}
\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*
Yes\, line 4 has error \- "sub" instead of "my"\.

Perl 5\.005\, 5\.6\.x and 5\.8\.0 was tested\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=high

Site configuration information for perl v5.8.0:

Configured by sem at Wed Jul 17 19:07:24 MSD 2002.

Summary of my perl5 (revision 5.0 version 8 subversion 0) configuration:
  Platform:
    osname=freebsd, osvers=4.6-release-p1, archname=i386-freebsd-thread-multi
    uname='freebsd avim.ciam.ru 4.6-release-p1 freebsd 4.6-release-p1 #0: fri jul 12 20:33:42 msd 2002 root@orion.ciam.ru:usrsrcsyscompileavim i386 '
    config_args=''
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='gcc', ccflags ='-DHAS_FPSETMASK -DHAS_FLOATINGPOINT_H -fno-strict-aliasing -I/usr/local/include',
    optimize='-O',
    cppflags='-DHAS_FPSETMASK -DHAS_FLOATINGPOINT_H -fno-strict-aliasing -I/usr/local/include'
    ccversion='', gccversion='2.95.3 20010315 (release) [FreeBSD]', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags ='-pthread -Wl,-E  -L/usr/local/lib'
    libpth=/usr/lib /usr/local/lib
    libs=-lgdbm -lm -lc_r -lcrypt -lutil
    perllibs=-lm -lc_r -lcrypt -lutil
    libc=, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags=' '
    cccdlflags='-DPIC -fpic', lddlflags='-shared  -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.8.0:
    /usr/local/lib/perl5/5.8.0/i386-freebsd-thread-multi
    /usr/local/lib/perl5/5.8.0
    /usr/local/lib/perl5/site_perl/5.8.0/i386-freebsd-thread-multi
    /usr/local/lib/perl5/site_perl/5.8.0
    /usr/local/lib/perl5/site_perl
    .


Environment for perl v5.8.0:
    HOME=/home/sem
    LANG=ru_RU.KOI8-R
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/sem/bin:/sbin:/usr/sbin:/bin:/usr/bin:/usr/local/bin:/usr/local/sbin:/usr/X11R6/bin:.
    PERL_BADLANG (unset)
    SHELL=/bin/tcsh


```
</details>

**Comment by p5pRT** [comment]
### From @nwc10

On Wed\, Jul 17\, 2002 at 06&#8203;:09&#8203;:11PM \-0000\, sem@&#8203;ciam\.ru \(via RT\) wrote&#8203;:

> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please enter your report here\]
> This code always make perl core dump&#8203;:
> \*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*
> my %level\_warnings;
> sub stat\_level\_warning
> \{
>     sub \($level\, $message\) = @&#8203;\_;
>     if\( $level\_warnings\{$message\} eq undef \)
>     \{
>     $level\_warnings\{$message\} = \[\];
>     for my $l \( 0 \.\. $levels \)
>     \{
>         $level\_warnings\{$message\}\->\[$l\] = 0;
>     \}
>     \}
>     $level\_warnings\{$message\}\->\[$level\]\+\+;
> \}
> \*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*
> Yes\, line 4 has error \- "sub" instead of "my"\.
> 
> Perl 5\.005\, 5\.6\.x and 5\.8\.0 was tested\.

>     osname=freebsd\, osvers=4\.6\-release\-p1\, archname=i386\-freebsd\-thread\-multi

I can't get your code to crash on FreeBSD 4\.6 with perl \(patchlevel 17521\,
my perl is not threaded\) but I can get it to crash on arm linux and x86
Linux\. On both I can reduce it to this&#8203;:

sub stat\_level\_warning
\{
&nbsp;    sub \(\);
&nbsp;    if\( $level\_warnings\{0\} \)
&nbsp;    \{
&nbsp;        for my $l \( 0 \.\. 0 \)
&nbsp;        \{
&nbsp;            $level\_warnings\->\[0\];
&nbsp;        \}
&nbsp;    \}
\}

Program received signal SIGSEGV\, Segmentation fault\.
Perl\_pad\_free \(po=4\) at op\.c&#8203;:611
611             SvPADTMP\_off\(PL\_curpad\[po\]\);
\(gdb\) where
\#0  Perl\_pad\_free \(po=4\) at op\.c&#8203;:611
\#1  0x00044c38 in Perl\_op\_clear \(o=0x124b90\) at op\.c&#8203;:917
\#2  0x00044c38 in Perl\_op\_clear \(o=0x124b90\) at op\.c&#8203;:917
\#3  0x00044b88 in Perl\_op\_free \(o=0x124b90\) at op\.c&#8203;:788
\#4  0x00044b48 in Perl\_op\_free \(o=0x124be8\) at op\.c&#8203;:776
\#5  0x00044b48 in Perl\_op\_free \(o=0x124c50\) at op\.c&#8203;:776
\#6  0x00044b48 in Perl\_op\_free \(o=0x124ce0\) at op\.c&#8203;:776
\#7  0x00044b48 in Perl\_op\_free \(o=0x124d08\) at op\.c&#8203;:776
\#8  0x00044b48 in Perl\_op\_free \(o=0x124d28\) at op\.c&#8203;:776
\#9  0x00044b48 in Perl\_op\_free \(o=0x124d90\) at op\.c&#8203;:776
\#10 0x00044b48 in Perl\_op\_free \(o=0x124db8\) at op\.c&#8203;:776
\#11 0x00044b48 in Perl\_op\_free \(o=0x124de0\) at op\.c&#8203;:776
\#12 0x00044b48 in Perl\_op\_free \(o=0x124e40\) at op\.c&#8203;:776
\#13 0x0004e164 in Perl\_newATTRSUB \(floor=107\, o=0x0\, proto=0x0\, attrs=0x0\, 
&nbsp;    block=0x124e40\) at op\.c&#8203;:5000
\#14 0x00042204 in Perl\_yyparse \(\) at perly\.y&#8203;:355
\#15 0x000201e8 in S\_parse\_body \(env=0x0\, xsinit=0x1dc7c \<xs\_init>\)
&nbsp;    at perl\.c&#8203;:1519
\#16 0x0001fca0 in perl\_parse \(my\_perl=0x4\, xsinit=0x1dc7c \<xs\_init>\, argc=3\, 
&nbsp;    argv=0xbffffd14\, env=0x0\) at perl\.c&#8203;:1041
\#17 0x0001dc08 in main \(argc=3\, argv=0xbffffd14\, env=0x6\) at perlmain\.c&#8203;:83
\#18 0x400e4b54 in \_\_libc\_start\_main \(\) from /lib/libc\.so\.6
\#19 0x400e4b54 in \_\_libc\_start\_main \(\) from /lib/libc\.so\.6
\#20 0x00080816 in Perl\_sv\_2pv\_flags \(sv=0x1ccb0\, lp=0x3\, flags=6\) at sv\.c&#8203;:3089
Cannot access memory at address 0x3003e59c
\(gdb\) print po
$4 = 4
\(gdb\) print PL\_curpad\[po\]
$5 = \(SV \*\) 0x6

PL\_curpad\[po\] was 0x19 on your version of the code\.

Running under valgrind I see this&#8203;:

nick@&#8203;penfold&#8203;:\~/build$ valgrind /usr/local/perl17559\-g/bin/perl5\.8\.0 crash
==29470== valgrind\-1\.0pre2\, a memory error detector for x86 GNU/Linux\.
==29470== Copyright \(C\) 2000\-2002\, and GNU GPL'd\, by Julian Seward\.
==29470== Estimated CPU clock rate is 262 MHz
==29470== For more details\, rerun with&#8203;: \-v
==29470==
==29470== Invalid read of size 4
==29470==    at 0x80894E8&#8203;: Perl\_pad\_free \(op\.c&#8203;:610\)
==29470==    by 0x8089A08&#8203;: Perl\_op\_clear \(op\.c&#8203;:917\)
==29470==    by 0x8089772&#8203;: Perl\_op\_free \(op\.c&#8203;:788\)
==29470==    by 0x8089715&#8203;: Perl\_op\_free \(op\.c&#8203;:776\)
==29470==    Address 0x40B8C2AC is 0 bytes after a block of size 16 alloc'd
==29470==    at 0x40042BDB&#8203;: malloc \(vg\_clientfuncs\.c&#8203;:100\)
==29470==    by 0x80AFB87&#8203;: Perl\_safesysmalloc \(util\.c&#8203;:78\)
==29470==    by 0x80C1477&#8203;: Perl\_av\_extend \(av\.c&#8203;:150\)
==29470==    by 0x80C1898&#8203;: Perl\_av\_store \(av\.c&#8203;:275\)
syntax error at crash line 3\, near "sub \(\);"
Execution of crash aborted due to compilation errors\.
==29470==
==29470== ERROR SUMMARY&#8203;: 1 errors from 1 contexts \(suppressed&#8203;: 0 from 0\)
==29470== malloc/free&#8203;: in use at exit&#8203;: 42374 bytes in 479 blocks\.
==29470== malloc/free&#8203;: 597 allocs\, 118 frees\, 49668 bytes allocated\.
==29470== For a detailed leak analysis\,  rerun with&#8203;: \-\-leak\-check=yes
==29470== For counts of detected errors\, rerun with&#8203;: \-v

\(that same perl and same script will SEGV when not running under valgrind\)

I don't know what to do next\, but I hope that this information is useful to
someone else\.

Nicholas Clark
\-\- 
Even better than the real thing&#8203;:    http&#8203;://nms\-cgi\.sourceforge\.net/

**Comment by p5pRT** [comment]
### From perlbug@perl.org

On Wed\, Jul 17\, 2002 at 06&#8203;:09&#8203;:11PM \-0000\, sem@&#8203;ciam\.ru \(via RT\) wrote&#8203;:

> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please enter your report here\]
> This code always make perl core dump&#8203;:
> \*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*
> my %level\_warnings;
> sub stat\_level\_warning
> \{
>     sub \($level\, $message\) = @&#8203;\_;
>     if\( $level\_warnings\{$message\} eq undef \)
>     \{
>     $level\_warnings\{$message\} = \[\];
>     for my $l \( 0 \.\. $levels \)
>     \{
>         $level\_warnings\{$message\}\->\[$l\] = 0;
>     \}
>     \}
>     $level\_warnings\{$message\}\->\[$level\]\+\+;
> \}
> \*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*
> Yes\, line 4 has error \- "sub" instead of "my"\.
> 
> Perl 5\.005\, 5\.6\.x and 5\.8\.0 was tested\.

>     osname=freebsd\, osvers=4\.6\-release\-p1\, archname=i386\-freebsd\-thread\-multi

I can't get your code to crash on FreeBSD 4\.6 with perl \(patchlevel 17521\,
my perl is not threaded\) but I can get it to crash on arm linux and x86
Linux\. On both I can reduce it to this&#8203;:

sub stat\_level\_warning
\{
&nbsp;    sub \(\);
&nbsp;    if\( $level\_warnings\{0\} \)
&nbsp;    \{
&nbsp;        for my $l \( 0 \.\. 0 \)
&nbsp;        \{
&nbsp;            $level\_warnings\->\[0\];
&nbsp;        \}
&nbsp;    \}
\}

Program received signal SIGSEGV\, Segmentation fault\.
Perl\_pad\_free \(po=4\) at op\.c&#8203;:611
611             SvPADTMP\_off\(PL\_curpad\[po\]\);
\(gdb\) where
\#0  Perl\_pad\_free \(po=4\) at op\.c&#8203;:611
\#1  0x00044c38 in Perl\_op\_clear \(o=0x124b90\) at op\.c&#8203;:917
\#2  0x00044c38 in Perl\_op\_clear \(o=0x124b90\) at op\.c&#8203;:917
\#3  0x00044b88 in Perl\_op\_free \(o=0x124b90\) at op\.c&#8203;:788
\#4  0x00044b48 in Perl\_op\_free \(o=0x124be8\) at op\.c&#8203;:776
\#5  0x00044b48 in Perl\_op\_free \(o=0x124c50\) at op\.c&#8203;:776
\#6  0x00044b48 in Perl\_op\_free \(o=0x124ce0\) at op\.c&#8203;:776
\#7  0x00044b48 in Perl\_op\_free \(o=0x124d08\) at op\.c&#8203;:776
\#8  0x00044b48 in Perl\_op\_free \(o=0x124d28\) at op\.c&#8203;:776
\#9  0x00044b48 in Perl\_op\_free \(o=0x124d90\) at op\.c&#8203;:776
\#10 0x00044b48 in Perl\_op\_free \(o=0x124db8\) at op\.c&#8203;:776
\#11 0x00044b48 in Perl\_op\_free \(o=0x124de0\) at op\.c&#8203;:776
\#12 0x00044b48 in Perl\_op\_free \(o=0x124e40\) at op\.c&#8203;:776
\#13 0x0004e164 in Perl\_newATTRSUB \(floor=107\, o=0x0\, proto=0x0\, attrs=0x0\, 
&nbsp;    block=0x124e40\) at op\.c&#8203;:5000
\#14 0x00042204 in Perl\_yyparse \(\) at perly\.y&#8203;:355
\#15 0x000201e8 in S\_parse\_body \(env=0x0\, xsinit=0x1dc7c \<xs\_init>\)
&nbsp;    at perl\.c&#8203;:1519
\#16 0x0001fca0 in perl\_parse \(my\_perl=0x4\, xsinit=0x1dc7c \<xs\_init>\, argc=3\, 
&nbsp;    argv=0xbffffd14\, env=0x0\) at perl\.c&#8203;:1041
\#17 0x0001dc08 in main \(argc=3\, argv=0xbffffd14\, env=0x6\) at perlmain\.c&#8203;:83
\#18 0x400e4b54 in \_\_libc\_start\_main \(\) from /lib/libc\.so\.6
\#19 0x400e4b54 in \_\_libc\_start\_main \(\) from /lib/libc\.so\.6
\#20 0x00080816 in Perl\_sv\_2pv\_flags \(sv=0x1ccb0\, lp=0x3\, flags=6\) at sv\.c&#8203;:3089
Cannot access memory at address 0x3003e59c
\(gdb\) print po
$4 = 4
\(gdb\) print PL\_curpad\[po\]
$5 = \(SV \*\) 0x6

PL\_curpad\[po\] was 0x19 on your version of the code\.

Running under valgrind I see this&#8203;:

nick@&#8203;penfold&#8203;:\~/build$ valgrind /usr/local/perl17559\-g/bin/perl5\.8\.0 crash
==29470== valgrind\-1\.0pre2\, a memory error detector for x86 GNU/Linux\.
==29470== Copyright \(C\) 2000\-2002\, and GNU GPL'd\, by Julian Seward\.
==29470== Estimated CPU clock rate is 262 MHz
==29470== For more details\, rerun with&#8203;: \-v
==29470==
==29470== Invalid read of size 4
==29470==    at 0x80894E8&#8203;: Perl\_pad\_free \(op\.c&#8203;:610\)
==29470==    by 0x8089A08&#8203;: Perl\_op\_clear \(op\.c&#8203;:917\)
==29470==    by 0x8089772&#8203;: Perl\_op\_free \(op\.c&#8203;:788\)
==29470==    by 0x8089715&#8203;: Perl\_op\_free \(op\.c&#8203;:776\)
==29470==    Address 0x40B8C2AC is 0 bytes after a block of size 16 alloc'd
==29470==    at 0x40042BDB&#8203;: malloc \(vg\_clientfuncs\.c&#8203;:100\)
==29470==    by 0x80AFB87&#8203;: Perl\_safesysmalloc \(util\.c&#8203;:78\)
==29470==    by 0x80C1477&#8203;: Perl\_av\_extend \(av\.c&#8203;:150\)
==29470==    by 0x80C1898&#8203;: Perl\_av\_store \(av\.c&#8203;:275\)
syntax error at crash line 3\, near "sub \(\);"
Execution of crash aborted due to compilation errors\.
==29470==
==29470== ERROR SUMMARY&#8203;: 1 errors from 1 contexts \(suppressed&#8203;: 0 from 0\)
==29470== malloc/free&#8203;: in use at exit&#8203;: 42374 bytes in 479 blocks\.
==29470== malloc/free&#8203;: 597 allocs\, 118 frees\, 49668 bytes allocated\.
==29470== For a detailed leak analysis\,  rerun with&#8203;: \-\-leak\-check=yes
==29470== For counts of detected errors\, rerun with&#8203;: \-v

\(that same perl and same script will SEGV when not running under valgrind\)

I don't know what to do next\, but I hope that this information is useful to
someone else\.

Nicholas Clark
\-\- 
Even better than the real thing&#8203;:    http&#8203;://nms\-cgi\.sourceforge\.net/

**Comment by p5pRT** [comment]
### From @floatingatoll

\[nick@&#8203;unfortu\.net \- Wed Jul 17 12&#8203;:55&#8203;:36 2002\]&#8203;:

> On Wed\, Jul 17\, 2002 at 06&#8203;:09&#8203;:11PM \-0000\, sem@&#8203;ciam\.ru \(via RT\) wrote&#8203;:

> sub stat\_level\_warning
> \{
>     sub \(\);
>     if\( $level\_warnings\{0\} \)
>     \{
>         for my $l \( 0 \.\. 0 \)
>         \{
>             $level\_warnings\->\[0\];
>         \}
>     \}
> \}

Reduced the test case to&#8203;:

\#\!perl
sub k \{
&nbsp;    sub \(\);
&nbsp;    for $l \( 0 \.\. 0 \) \{ $m \}
\}
\_\_END\_\_

Running perl with \-Dl gives you some interesting output&#8203;:

> \./perl \-Dl 15005\.pl
\(15005\.pl&#8203;:0\)    ENTER scope 2 at perly\.c&#8203;:1528
zsh&#8203;: segmentation fault \(core dumped\)  \./perl \-Dl 15005\.pl

Included also are a gdb and valgrind against @&#8203;17821\, redhat x86\, ithreads\.

R\.

===> gdb stack trace\, perl@&#8203;17821

Program received signal SIGSEGV\, Segmentation fault\.
0x187371 in Perl\_pad\_swipe \(my\_perl=0x804c330\, po=4\) at op\.c&#8203;:635
635         SvPADTMP\_off\(PL\_curpad\[po\]\);
\(gdb\) bt
\#0  0x187371 in Perl\_pad\_swipe \(my\_perl=0x804c330\, po=4\) at op\.c&#8203;:635
\#1  0x187710 in Perl\_op\_clear \(my\_perl=0x804c330\, o=0x8066f48\) at op\.c&#8203;:822
\#2  0x1875c5 in Perl\_op\_free \(my\_perl=0x804c330\, o=0x8066f48\) at op\.c&#8203;:788
\#3  0x187566 in Perl\_op\_free \(my\_perl=0x804c330\, o=0x8067020\) at op\.c&#8203;:776
\#4  0x187566 in Perl\_op\_free \(my\_perl=0x804c330\, o=0x80670b8\) at op\.c&#8203;:776
\#5  0x187566 in Perl\_op\_free \(my\_perl=0x804c330\, o=0x8067220\) at op\.c&#8203;:776
\#6  0x187566 in Perl\_op\_free \(my\_perl=0x804c330\, o=0x8067248\) at op\.c&#8203;:776
\#7  0x187566 in Perl\_op\_free \(my\_perl=0x804c330\, o=0x8067268\) at op\.c&#8203;:776
\#8  0x187566 in Perl\_op\_free \(my\_perl=0x804c330\, o=0x80672f0\) at op\.c&#8203;:776
\#9  0x19635a in Perl\_newATTRSUB \(my\_perl=0x804c330\, floor=107\, o=
0x8066e50\, 
&nbsp;    proto=0x0\, attrs=0x0\, block=0x80672f0\) at op\.c&#8203;:4992
\#10 0x183273 in Perl\_yyparse \(my\_perl=0x804c330\) at perly\.y&#8203;:355
\#11 0x1495cb in S\_parse\_body \(my\_perl=0x804c330\, env=0x0\, 
&nbsp;    xsinit=0x80494a0 \<xs\_init>\) at perl\.c&#8203;:1521
\#12 0x148171 in perl\_parse \(my\_perl=0x804c330\, xsinit=0x80494a0 \<xs\_init>\, 
&nbsp;    argc=2\, argv=0xbffff494\, env=0x0\) at perl\.c&#8203;:1043
\#13 0x804940d in main \(argc=2\, argv=0xbffff494\, env=0xbffff4a0\)
&nbsp;    at perlmain\.c&#8203;:83

===> valgrind 1\.0\.1 output

==32201== valgrind\-1\.0\.1\, a memory error detector for x86 GNU/Linux\.
==32201== Copyright \(C\) 2000\-2002\, and GNU GPL'd\, by Julian Seward\.
==32201== Estimated CPU clock rate is 746 MHz
==32201== For more details\, rerun with&#8203;: \-v
==32201== 
==32201== pthread\_mutex\_unlock&#8203;: mutex is not locked
==32201==    at 0x540D87&#8203;: \_\_pthread\_mutex\_unlock \(in /usr/local/lib/valgrind/
libpthread\.so\)
==32201==    by 0x62F20C&#8203;: \_\_register\_frame\_info \(in /lib/libc\-2\.1\.3\.so\)
==32201==    by 0x672FF0&#8203;: \(within /lib/libutil\-2\.1\.3\.so\)
==32201==    by 0x672CD1&#8203;: \(within /lib/libutil\-2\.1\.3\.so\)
==32201== 
==32201== pthread\_mutex\_destroy&#8203;: mutex is still in use
==32201==    at 0x540004&#8203;: pthread\_error \(vg\_libpthread\.c&#8203;:233\)
==32201==    by 0x540DEA&#8203;: \_\_pthread\_mutex\_destroy \(in /usr/local/lib/
valgrind/libpthread\.so\)
==32201==    by 0x5A5FE5&#8203;: \_IO\_default\_finish \(genops\.c&#8203;:592\)
==32201==    by 0x5A5636&#8203;: \_IO\_file\_finish@&#8203;@&#8203;GLIBC\_2\.1 \(fileops\.c&#8203;:163\)
==32201== 
==32201== Invalid read of size 4
==32201==    at 0x3AA6F7&#8203;: Perl\_op\_clear \(in /home/rs/build/perl\-current/
libperl\.so\)
==32201==    by 0x3AA5C5&#8203;: Perl\_op\_free \(in /home/rs/build/perl\-current/
libperl\.so\)
==32201==    by 0x3AA566&#8203;: Perl\_op\_free \(in /home/rs/build/perl\-current/
libperl\.so\)
==32201==    by 0x3AA566&#8203;: Perl\_op\_free \(in /home/rs/build/perl\-current/
libperl\.so\)
==32201==    Address 0x101D928 is 0 bytes after a block of size 16 alloc'd
==32201==    at 0x153DD9&#8203;: malloc \(in /usr/local/lib/valgrind/valgrind\.so\)
==32201==    by 0x3E2CD6&#8203;: Perl\_safesysmalloc \(in /home/rs/build/perl\-current/
libperl\.so\)
==32201==    by 0x3FE4EF&#8203;: Perl\_av\_extend \(in /home/rs/build/perl\-current/
libperl\.so\)
==32201==    by 0x3FF190&#8203;: Perl\_av\_store \(in /home/rs/build/perl\-current/
libperl\.so\)
==32201== 
==32201== Invalid read of size 4
==32201==    at 0x3AA358&#8203;: Perl\_pad\_swipe \(in /home/rs/build/perl\-current/
libperl\.so\)
==32201==    by 0x3AA710&#8203;: Perl\_op\_clear \(in /home/rs/build/perl\-current/
libperl\.so\)
==32201==    by 0x3AA5C5&#8203;: Perl\_op\_free \(in /home/rs/build/perl\-current/
libperl\.so\)
==32201==    by 0x3AA566&#8203;: Perl\_op\_free \(in /home/rs/build/perl\-current/
libperl\.so\)
==32201==    Address 0x101D928 is 0 bytes after a block of size 16 alloc'd
==32201==    at 0x153DD9&#8203;: malloc \(in /usr/local/lib/valgrind/valgrind\.so\)
==32201==    by 0x3E2CD6&#8203;: Perl\_safesysmalloc \(in /home/rs/build/perl\-current/
libperl\.so\)
==32201==    by 0x3FE4EF&#8203;: Perl\_av\_extend \(in /home/rs/build/perl\-current/
libperl\.so\)
==32201==    by 0x3FF190&#8203;: Perl\_av\_store \(in /home/rs/build/perl\-current/
libperl\.so\)
==32201== 
==32201== Invalid read of size 4
==32201==    at 0x3AA36B&#8203;: Perl\_pad\_swipe \(in /home/rs/build/perl\-current/
libperl\.so\)
==32201==    by 0x3AA710&#8203;: Perl\_op\_clear \(in /home/rs/build/perl\-current/
libperl\.so\)
==32201==    by 0x3AA5C5&#8203;: Perl\_op\_free \(in /home/rs/build/perl\-current/
libperl\.so\)
==32201==    by 0x3AA566&#8203;: Perl\_op\_free \(in /home/rs/build/perl\-current/
libperl\.so\)
==32201==    Address 0x101D928 is 0 bytes after a block of size 16 alloc'd
==32201==    at 0x153DD9&#8203;: malloc \(in /usr/local/lib/valgrind/valgrind\.so\)
==32201==    by 0x3E2CD6&#8203;: Perl\_safesysmalloc \(in /home/rs/build/perl\-current/
libperl\.so\)
==32201==    by 0x3FE4EF&#8203;: Perl\_av\_extend \(in /home/rs/build/perl\-current/
libperl\.so\)
==32201==    by 0x3FF190&#8203;: Perl\_av\_store \(in /home/rs/build/perl\-current/
libperl\.so\)
==32201== 
==32201== Invalid read of size 4
==32201==    at 0x3AA371&#8203;: Perl\_pad\_swipe \(in /home/rs/build/perl\-current/
libperl\.so\)
==32201==    by 0x3AA710&#8203;: Perl\_op\_clear \(in /home/rs/build/perl\-current/
libperl\.so\)
==32201==    by 0x3AA5C5&#8203;: Perl\_op\_free \(in /home/rs/build/perl\-current/
libperl\.so\)
==32201==    by 0x3AA566&#8203;: Perl\_op\_free \(in /home/rs/build/perl\-current/
libperl\.so\)
==32201==    Address 0x8 is not stack'd\, malloc'd or free'd

===> \./perl \-Dl 15005\.pl
\(15005\.pl&#8203;:0\)    ENTER scope 2 at perly\.c&#8203;:1528
zsh&#8203;: segmentation fault \(core dumped\)  \./perl \-Dl 15005\.pl

**Comment by p5pRT** [comment]
### From spamme@gameknot.com

Here's another example of the code that causes the compiler to seg fault
\(core dumped\)&#8203;:

\#\!/usr/bin/perl             
&nbsp;                            
use strict;                 
&nbsp;                            
my $pooh = pooh\( "???"\, 1 \);
&nbsp;                            
sub pooh                    
\{                           
&nbsp;  sub \($result\,$swap\) = @&#8203;\_; 
&nbsp;                            
&nbsp;  my $r1 = 0;               
&nbsp;  my $r2 = 0;               
&nbsp;                            
&nbsp;  return "$r1\-$r2";         
\}

**Comment by p5pRT** [comment]
### From @jhi

Perl has now been patched to give the following error&#8203;:
Illegal declaration of anonymous subroutine at x line 4\.
The fix will be in Perl 5\.8\.1\, whenever that happens\.
I'm marking the problem ticket as resolved\.

**Comment by p5pRT** [comment]
@jhi - Status changed from 'new' to 'resolved'
